python/pdstools/reports/ModelReport.qmd

```diff
@@ -197,65 +328,135 @@ predictors. This can be configured via a parameter to this report.
+
+    # Add line for base propensity, TODO consider putting back in the library
+    fig.add_hline(y=base_propensity, line_dash="dash", line_color="grey", yref='y2')
+    # fig.update_xaxes(type='category') # prevent plotly from trying to guess
+
+    fig.update_layout(width=700, height=300)
+    fig.show()
+
+    # "Philip Mann" plot with simple red/green lift bars relative to base propensity
+    # TODO move below "Philip Mann" plot into a plot function (plotBinningLift)
+
+    pm_plot_binning_table = (
+        predictor_data
+        .select(
+            pl.col(["BinSymbol", "BinPositives", "BinIndex"]), 
+            (lift(pl.col("BinPositives"), pl.col("BinNegatives")) - 1.0)
+        )
+        .with_columns(
+            pl.when((pl.col("Lift") >= 0.0) & (pl.col("BinPositives") > 5)).then(pl.lit('pos'))
+            .when((pl.col("Lift") >= 0.0) & (pl.col("BinPositives") <= 5)).then(pl.lit('pos_shaded'))
+            .when((pl.col("Lift") < 0.0) & (pl.col("BinPositives") > 5)).then(pl.lit('neg'))
+            .otherwise(pl.lit('neg_shaded'))
+            .alias("Direction"))
+        .sort("BinIndex")
+        .collect()
+    )
+
+    fig = px.bar(
+        x=pm_plot_binning_table['Lift'],
+        y=pm_plot_binning_table['BinSymbol'], 
+        color=pm_plot_binning_table['Direction'],
+        color_discrete_map = {
+            'neg': '#A01503', 
+            'pos': '#5F9F37',
+            'neg_shaded' : '#DAA9AB',
+            'pos_shaded' : '#C5D9B7'},
+        orientation='h',
+        template='pega'
+    )
+    fig.update_yaxes(type='category')
+    fig.layout.xaxis.tickformat = ",.2%"
+    fig.add_vline(x=0, line_color="black")
+    fig.update_layout(
+        yaxis = dict(autorange="reversed"),
+        title="Propensity Lift", xaxis_title="", yaxis_title="")
+    # Not convinced why we need below. It should respect the binindex order but w/o this it doesnt always.
+    fig.update_yaxes(categoryorder='array', categoryarray=pm_plot_binning_table['BinSymbol'])
+    # TODO: when there are many bins not all are always shown, figure out how to fix
+
+    fig.update_layout(width=700, height=300, showlegend=False)
     fig.show()
 
+    # Table with binning details
+
+    human_friendly_binning_table = (
+        predictor_data
+        .select(
+            pl.col("BinIndex").alias("Index"),
+            pl.col("BinSymbol").alias("Bin"),
+            pl.col("BinResponseCount").alias("Responses"),
+            pl.col("BinPositives").alias("Positives"),
+            zRatio(pl.col("BinPositives"), pl.col("BinNegatives")).round(3).alias("Z Ratio"),
+            (lift(pl.col("BinPositives"), pl.col("BinNegatives")) * 100).round(2).alias("Lift (%)"),
+            (pl.col("BinPropensity") * 100).round(3).alias("Propensity (%)")
+        )
+        .collect()
+    )
+
+    display(Markdown(
+        tabulate([[f[i] for f in human_friendly_binning_table] for i in range(human_friendly_binning_table.shape[0])],
```

Comment:
Why this odd comprehension? Can't you just do `[i for i in range(human_friendly_binning_table)]`?

Reply:
I need a transposed list of lists